packages/suite/src/components/firmware/SelectCustomFirmware.tsx

```diff
@@ -66,15 +69,24 @@ const Step: React.FC<StepProps> = ({ order, title, children }) => (
     </StepContainer>
 );
 
-export const SelectCustomFirmware = ({ onSuccess }: { onSuccess: (fw: ArrayBuffer) => void }) => {
+type Props = {
+    device?: TrezorDevice;
+    onSuccess: (fw: ArrayBuffer) => void;
+};
+
+export const SelectCustomFirmware = ({ device, onSuccess }: Props) => {
     const [firmwareBinary, setFirmwareBinary] = useState<ArrayBuffer>();
+    const { translationString } = useTranslation();
 
-    const onFirmwareUpload = (firmware: File, setError: (msg: string) => void) => {
+    const onFirmwareUpload = async (firmware: File, setError: (msg: string) => void) => {
         if (!firmware?.name?.endsWith('.bin')) {
             setError('file-type');
             return;
         }
-        firmware.arrayBuffer().then(setFirmwareBinary);
+        const fw = await firmware.arrayBuffer();
+        const validationError = validateFirmware(fw, device);
+        if (validationError) setError(translationString(validationError));
+        else setFirmwareBinary(fw);
```

Comment:
nit: I think that we do not use this syntax in the project.
```suggestion
    if (validationError) {
        setError(translationString(validationError));
    } else {
        setFirmwareBinary(fw);
    }
```

Reply:
Done.